File: basic_dsa_two_hundred_nine.cpp
/**
 * Basic DSA - Minimum Size Subarray Sum (LeetCode 209)
 * Find minimal length subarray with sum >= target (sliding window).
 */
#include <iostream>
#include <vector>
#include <algorithm>
#include <climits>

int minSubArrayLen(int target, std::vector<int>& nums) {
    int ans = INT_MAX, sum = 0, lo = 0;
    for (int hi = 0; hi < (int)nums.size(); hi++) {
        sum += nums[hi];
        while (sum >= target) {
            ans = std::min(ans, hi - lo + 1);
            sum -= nums[lo++];
        }
    }
    return ans == INT_MAX ? 0 : ans;
}

int main() {
    std::cout << std::string(60, '=') << "\n";
    std::cout << "Problem: Minimum Size Subarray Sum\n";
    std::cout << std::string(60, '=') << "\n";

    std::vector<int> a1 = {2, 3, 1, 2, 4, 3};
    std::cout << "target=7, [2,3,1,2,4,3] -> " << minSubArrayLen(7, a1) << "\n";

    std::vector<int> a2 = {1, 4, 4};
    std::cout << "target=4, [1,4,4] -> " << minSubArrayLen(4, a2) << "\n";

    std::vector<int> a3 = {1, 1, 1, 1, 1, 1, 1, 1};
    std::cout << "target=11, [1,1,1,1,1,1,1,1] -> " << minSubArrayLen(11, a3) << "\n";
}

File: basic_dsa_two_hundred_nine.py
"""
Basic DSA - Find All Anagrams in a String
Given two strings s and p, return a list of all start indices of p's anagrams in s.
The answer can be in any order.
"""


def find_anagrams(s, p):
    """
    Sliding window character counts over lowercase English letters.
    Time Complexity: O(n)
    Space Complexity: O(1)
    """
    if len(p) > len(s):
        return []

    def idx(ch):
        return ord(ch) - ord("a")

    need = [0] * 26
    win = [0] * 26
    for ch in p:
        need[idx(ch)] += 1

    res = []
    m = len(p)
    for i, ch in enumerate(s):
        win[idx(ch)] += 1
        if i >= m:
            win[idx(s[i - m])] -= 1
        if i >= m - 1 and win == need:
            res.append(i - m + 1)
    return res


if __name__ == "__main__":
    print("=" * 60)
    print("Problem: Find All Anagrams in a String")
    print("=" * 60)

    print(find_anagrams("cbaebabacd", "abc"))  # [0, 6]
    print(find_anagrams("abab", "ab"))  # [0, 1, 2]
